syntax: support inline comments inside backquotes

For long commands and pipelines split across multiple lines, a common workaround to add comments in the middle is to use the following syntax, slightly abusing subshells, but seems to be the cleanest way overall:

```
command -option value `# comment goes here` \
  -another-option another-value `# another comment goes here` \
  -xyz
```

This syntax seems to be explicitly supported by `shellcheck`, but `shfmt` either gets confused or replaces it with `$(...)` subshell syntax (which won't generally work properly in combination with `#`). This in turn triggers `shellcheck` warnings. I think you can use `$(: ...)` as a workaround, but then you also have to disable the `shellcheck` warning for an unquoted substitution (quoting it would result in extraneous empty parameters).

Would it be possible for `shfmt` to recognize this special comment syntax and leave it alone (possibly under the control of a command-line flag)?

Thanks, I've read that and it looks applicable here. It doesn't resolve what to do with existing code that looks like this though:

```
  # Use aquery to get the list of output files of the `CreatePatch` action,
  # Then strip the patch path down to that of its source file, and apply
  # the patch file generated by Bazel to the original source file.
  # shellcheck disable=SC2016
  CLANG_FORMAT="$(clangFormatLocation)" runBazel aquery --config lint --include_aspects --output_groups clang_format_patches_only "mnemonic(\"CreatePatch\", $(generateAqueryTargetString))" --output textproto \
    `# Get relative paths to source files` \
    `# perl used instead of grep --perl-regexp since grep macOS doesnt support it` \
    | perl -ne "while(/(?<=exec_path: \"${PREFIX//\//\\/}\/).*\.patch_.+(?=\")/g){print \"\$&\n\";}" \
    `# Create the patch commands which, when executed patch the source files.` \
    `# --binary flag used to correctly handle line endings on Windows.` \
    | xargs -L1 -I 'PATH_TO_SOURCE_FILE' echo 'FILE_PATH="PATH_TO_SOURCE_FILE"; echo patch --binary "${FILE_PATH%.patch_*}" "'"${BAZEL_BIN}"'/${FILE_PATH}"' \
    `# De-duplicate the patch commands, such that there is at most one patch command for each source file.` \
    `# There are N patch files per source files, where N is the number of bazel targets directly including the target.` \
    `# The format of the commands being de-duplicated is: patch --binary -source file- -patch file-` \
     | sh `# Calls the echo command generated by xargs, resulting in the patch command to be run by the next sh invocation below`\
     | sort --unique --field-separator=" " --key=3,3  `# Remove duplicate patch commands to the same source file` \
     | sh `# Execute patch commands`
```

I'm not sure of a workaround that still enables inline comments in long pipelines like above.

Moving all the comments to above the command would remove the valuable context of which part they're referring to.

--- a/syntax/filetests_test.go
+++ b/syntax/filetests_test.go
@@ -4492,6 +4492,25 @@ var fileTestsKeepComments = []testCase{
 			}},
 		},
 	},
+	{
+		Strs: []string{
+			"$(\n\t# foo\n)",
+			"`\n\t# foo\n`",
+			"`# foo\n`",
+		},
+		common: &CmdSubst{
+			Last: []Comment{{Text: " foo"}},
+		},
+	},
+	{
+		Strs: []string{
+			"`# foo`",
+			"` # foo`",
+		},
+		common: &CmdSubst{
+			Last: []Comment{{Text: " foo"}},
+		},
+	},
 }
 
 func fullProg(v interface{}) *File {

--- a/syntax/lexer.go
+++ b/syntax/lexer.go
@@ -267,10 +267,18 @@ skipSpace:
 		case '#':
 			r = p.rune()
 			p.newLit(r)
-			for r != '\n' && r != utf8.RuneSelf {
-				if r == escNewl {
+		runeLoop:
+			for {
+				switch r {
+				case '\n', utf8.RuneSelf:
+					break runeLoop
+				case escNewl:
 					p.litBs = append(p.litBs, '\\', '\n')
-					break
+					break runeLoop
+				case '`':
+					if p.backquoteEnd() {
+						break runeLoop
+					}
 				}
 				r = p.rune()
 			}

--- a/syntax/printer.go
+++ b/syntax/printer.go
@@ -627,6 +627,12 @@ func (p *Printer) wordPart(wp, next WordPart) {
 			p.nestedStmts(x.Stmts, x.Last, x.Right)
 			p.wantSpace = false
 			p.semiRsrv("}", x.Right)
+		// Special case: `# inline comment`
+		case x.Backquotes && len(x.Stmts) == 0 &&
+			len(x.Last) == 1 && x.Right.Line() == p.line:
+			p.WriteString("`#")
+			p.WriteString(x.Last[0].Text)
+			p.WriteString("`")
 		default:
 			p.WriteString("$(")
 			p.wantSpace = len(x.Stmts) > 0 && startsWithLparen(x.Stmts[0])

--- a/syntax/printer_test.go
+++ b/syntax/printer_test.go
@@ -20,7 +20,7 @@ func TestPrintCompact(t *testing.T) {
 	parserMirBSD := NewParser(KeepComments(true), Variant(LangMirBSDKorn))
 	parserBats := NewParser(KeepComments(true), Variant(LangBats))
 	printer := NewPrinter()
-	for _, c := range fileTests {
+	for _, c := range append(fileTests, fileTestsKeepComments...) {
 		t.Run("", func(t *testing.T) {
 			in := c.Strs[0]
 			parser := parserPosix
@@ -239,6 +239,7 @@ var printTests = []printCase{
 	samePrint("#before\nfoo && bar"),
 	samePrint("foo | bar # inline"),
 	samePrint("foo && bar # inline"),
+	samePrint("foo `# inline` \\\n\tbar"),
 	samePrint("for a in 1 2; do\n\n\tbar\ndone"),
 	{
 		"a \\\n\t&& b",
@@ -1078,7 +1079,7 @@ func TestPrintOptionsNotBroken(t *testing.T) {
 		{"SingleLine", []PrinterOption{SingleLine(true)}},
 	} {
 		printer := NewPrinter(opts.list...)
-		for i, tc := range fileTests {
+		for i, tc := range append(fileTests, fileTestsNoPrint...) {
 			t.Run(fmt.Sprintf("File%s%03d", opts.name, i), func(t *testing.T) {
 				parser := parserPosix
 				if tc.Bats != nil {

